### PR TITLE
memtest86-efi: compute ESP offset from partitions

### PR DESCRIPTION
**Pull request by emilazy**
###### Motivation for this change

Uses `sfdisk(1)` + `jq(1)` rather than hardcoding a partition offset, and avoids an intermediate `dd(1)` step by passing the offset directly to mtools.

Hopefully the derivation is less obscure / better documented now.

cc @flokli @cdepillabout

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by emilazy** [comment]
Follow-up to #86470.

**Review by cdepillabout — APPROVED** [review]
@emilazy Thanks for sending a PR updating this.

I think this is looking really good.  It is easy to understand what is going on here, and I feel comfortable keeping this updated in the future.

--------------------------------------

I have two small concerns, but I don't think they are big enough to prevent this from being merged in.

Like I said in https://github.com/NixOS/nixpkgs/pull/86470#discussion_r419028309, it looks like we are still potentially passing both partitions to `mtool`, and relying on `mtool` to figure out it should disregard the second one.  Although in this case, we could probably use a similar `sfdisk` call to figure out the length of the partition we want, and pull that out with `dd`, somewhat similar to what was done in https://github.com/NixOS/nixpkgs/pull/86470.  I'm okay with not doing that here in this PR, and making that change in the future if it becomes a problem.

My second small concern is that it appears the output format of `sfdisk` changes slightly from version to version?  I was playing around with `sfdisk`, and I was running into a weird problem of it not working.  I realized I was using `sfdisk` from 19.09.  The output format of the JSON is slightly different.  It didn't contain the `sectorsize` key.  This also isn't a big problem, and shouldn't prevent this from being merged in, but when I go to update this derivation, if it starts failing I'll have to keep this in mind as a possible cause.

**Comment by emilazy** [comment]
> Like I said in [#86470 (comment)](https://github.com/NixOS/nixpkgs/pull/86470#discussion_r419028309), it looks like we are still potentially passing both partitions to `mtool`, and relying on `mtool` to figure out it should disregard the second one. Although in this case, we could probably use a similar `sfdisk` call to figure out the length of the partition we want, and pull that out with `dd`, somewhat similar to what was done in #86470. I'm okay with not doing that here in this PR, and making that change in the future if it becomes a problem.

Specifying the location of partitions is the intended use of the offset functionality (see e.g. [the mtools manual](https://www.gnu.org/software/mtools/manual/mtools.html#location-information)), and in general is just how partitions and filesystems work. Filesystems generally encode their own extent, and there's no "dynamically-sized reads/writes" that would go outside their limits, so unless the image was maliciously constructed as invalid an FAT filesystem that points to sectors outside of the partition, there's no way mtools would even be aware of the rest of the image; that's presumably why it only takes an offset value to begin with, not an offset and a size. If the risk you were worrying about was real, then something like growing a partition would be much more risky and difficult, rather than the routine operation it is.

I don't claim that any of this is particularly obvious, but I think the fact that mtools accepts this syntax in the first place is suggestive of the fact that it's how things are meant to work, and I don't think a diversion about how partition tables and filesystem headers work would really be appropriate for this derivation. 
> My second small concern is that it appears the output format of `sfdisk` changes slightly from version to version? I was playing around with `sfdisk`, and I was running into a weird problem of it not working. I realized I was using `sfdisk` from 19.09. The output format of the JSON is slightly different. It didn't contain the `sectorsize` key. This also isn't a big problem, and shouldn't prevent this from being merged in, but when I go to update this derivation, if it starts failing I'll have to keep this in mind as a possible cause.

I guess the JSON API is probably pretty new; it was the least kludgy way I could find to parse GPT from userspace programmatically. If anyone knows a better alternative I'd be happy to use that instead.

**Comment by cdepillabout** [comment]
Hmm, I've discovered something interesting.

If I go back to before https://github.com/NixOS/nixpkgs/pull/86470, the files output are slightly different:

```console
$ cd /some/path/to/nixpkgs/
$ git checkout b7f80f00eff  # an older commit from before https://github.com/NixOS/nixpkgs/pull/86470 was merged in
$ nix-build -A memtest86-efi
$ ls result/
Benchmark  blacklist.cfg  BOOTIA32.efi  BOOTX64.efi  mt86.png  unifont.bin
$ git checkout 15172cf # the commit from this PR
$ nix-build -A memtest86-efi
$ ls result/
blacklist.cfg  BOOTIA32.efi  BOOTX64.efi  mt86.png  unifont.bin
```

It looks like the `Benchmark/` directory is no longer being pulled out from the partition by `mtool`.  This seems to be the case for both this PR (#86623) and the previous PR from @ajs124 (#86470).

After playing around with this for a while, I figured out that passing the `-s` flag to `mcopy` makes it successfully create the `Benchmark/` directory.

File: pkgs/tools/misc/memtest86-efi/default.nix [hunk 1]
```diff
@@ -22,19 +28,28 @@ stdenv.mkDerivation rec {
...
+    # the filesystem so that it can be installed directly on the hard drive.
+    IMG=$src/memtest86-usb.img
+    ESP_OFFSET=$(sfdisk --json $IMG | jq -r '
+      # Partition type GUID identifying EFI System Partitions
+      def ESP_GUID: "C12A7328-F81F-11D2-BA4B-00A0C93EC93B";
+      .partitiontable |
+      .sectorsize * (.partitions[] | select(.type == ESP_GUID) | .start)
+    ')
+    mkdir $out
+    mcopy -i $IMG@@$ESP_OFFSET ::'/EFI/BOOT/*' $out/
```
Comment (cdepillabout) [inline pkgs/tools/misc/memtest86-efi/default.nix:52]:
```suggestion
    mcopy -s -i $IMG@@$ESP_OFFSET ::'/EFI/BOOT/*' $out/
```

**Comment by cdepillabout** [comment]
Like I said above, this LGTM, but I'll give @ajs124 a few days to do a review (since @ajs124 is the author of the original PR).

If @ajs124 doesn't respond in a few days, @emilazy please feel free to ping me and I will merge this in.

File: pkgs/tools/misc/memtest86-efi/default.nix (hunk 1 above)
Comment (emilazy) [reply pkgs/tools/misc/memtest86-efi/default.nix]: Good catch; the directory is empty so hopefully this wouldn't affect behaviour in practice, but I added `-s` and `-v` (which will hopefully help flag future discrepancies up in the build logs).

**Review by ajs124 — APPROVED** [review]

**Comment by cdepillabout** [comment]
LGTM!

@emilazy Thanks for sending this PR, and @ajs124 thanks for reviewing this.

**Pull request merged**